pkg/defaults/defaults.go

```diff
@@ -144,8 +144,12 @@ func processOpSrc(client wrapper.Client, defaultOpsrc v1.OperatorSource) error {
 		return err
 	}
 
-	// Create if not present or is deleted
-	if errors.IsNotFound(err) || !opsrcCluster.ObjectMeta.DeletionTimestamp.IsZero() {
+	// Create if not present OR the object exists AND it was cleaned up already
+	// (marked for deletion, finalizer removed). However if it was marked for deletion
+	// and the garbage collector hasn't gotten around to it yet (i.e. the finalizers have
+	// not been removed) do nothing here and allow the deletion reconciler to remove the
+	// finalizer and recreate the object.
```

Comment:
We are waiting for the `finalizer` to be removed, then can we not wait until it is deleted by GC? Are there some edge cases with the deletion logic in marketplace or with the GC we are trying to work around here?

Reply:
It's not going to be removed because the object is marked for deletion but the operator isn't running, so the delete reconciler is never called and thus the finalizer won't be removed.

Reply:
okay, `len(opsrcCluster.Finalizers) == 0` is always going to be `true` since `processOpSrc` is always called after `out.RemoveFinalizer()`. https://github.com/operator-framework/operator-marketplace/blob/master/pkg/operatorsource/deleted.go#L77

Maybe that's why you need it? - https://github.com/operator-framework/operator-marketplace/blob/0d9a16a00b97d195487e4f4f9f244e635a19cfe3/pkg/operatorsource/succeeded.go#L83

Just a thought, maybe this is a better location to put the `defaults.Ensure()` call - https://github.com/operator-framework/operator-marketplace/blob/0d9a16a00b97d195487e4f4f9f244e635a19cfe3/pkg/operatorsource/outofsync.go#L63? This ensures that we are always in sync irrespective of the phase.

Please don't bother with my suggestion if your changes works :), I just wanted to understand the issue.